tests/adapter/dbt/tests/adapter/constraints/test_constraints.py

```diff
@@ -0,0 +1,159 @@
+
+        assert constraints_enabled_actual_config is True
+
+        expected_compile_error = "Please ensure the name, order, and number of columns in your `yml` file match the columns in your SQL file."
+        expected_schema_file_columns = "Schema File Columns: ['ID', 'COLOR', 'DATE_DAY']"
+        expected_sql_file_columns = "SQL File Columns: ['ERROR', 'COLOR', 'DATE_DAY']"
+
+        assert expected_compile_error in log_output
+        assert expected_schema_file_columns in log_output
+        assert expected_sql_file_columns in log_output
+
+
+# This is SUPER specific to Postgres, and will need replacing on other adapters
+# TODO: make more generic
+_expected_sql = """
+create table "{0}"."{1}"."my_model__dbt_tmp" (
+    id integer not null primary key check (id > 0) ,
+    color text ,
+    date_day date
+) ;
+insert into "{0}"."{1}"."my_model__dbt_tmp" (
+    id ,
+    color ,
+    date_day
+) (
+    select 1 as id,
+    'blue' as color,
+    cast('2019-01-01' as date) as date_day
+);
+"""
+
+
+class BaseConstraintsRuntimeEnforcement:
+    """
+    These constraints pass muster for dbt's preflight checks. Make sure they're
+    passed into the DDL statement. If they don't match up with the underlying data,
+    the data platform should raise an error at runtime.
+    """
+
+    @pytest.fixture(scope="class")
+    def models(self):
+        return {
+            "my_model.sql": my_model_sql,
+            "constraints_schema.yml": model_schema_yml,
+        }
+
+    @pytest.fixture(scope="class")
+    def expected_sql(self, project):
+        return _expected_sql.format(project.database, project.test_schema)
+
+    def test__constraints_ddl(self, project, expected_sql):
+        results = run_dbt(["run", "-s", "my_model"])
+        assert len(results) == 1
+        # TODO: consider refactoring this to introspect logs instead
+        with open("./target/run/test/models/my_model.sql", "r") as fp:
+            generated_sql = fp.read()
+
+        generated_sql_check = re.sub(r"\s+", " ", generated_sql).lower().strip()
+        expected_sql_check = re.sub(r"\s+", " ", expected_sql).lower().strip()
+        assert (
+            expected_sql_check == generated_sql_check
+        ), f"generated sql did not match expected: \n{generated_sql} \n{expected_sql}"
+
+    def test__constraints_enforcement_rollback(self, project, unique_schema):
+        results = run_dbt(["run", "-s", "my_model"])
+        assert len(results) == 1
+
+        # Make a contract-breaking change to the model
+        my_model_with_nulls = my_model_sql.replace("1 as id", "null as id")
+        write_file(my_model_with_nulls, "models", "my_model.sql")
+
+        failing_results = run_dbt(["run", "-s", "my_model"], expect_pass=False)
+        assert len(failing_results) == 1
+
+        # Verify the previous table still exists
+        old_model_exists_sql = """
+            select id from dbt.{0}.my_model where id = 1
```

Comment:
@jtcohen6 I think this will also need the templated database!

Reply:
done